Add test case for #548

puxlit opened the pull request:
Per https://github.com/pypa/setuptools_scm/pull/681#issuecomment-1031365308, this PR adds a test case for #548. This test case is split into its own PR to demonstrate that it fails without #681.

<details><summary>Test failing</summary>

```
(venv) ubuntu@ephemeral:~/setuptools_scm$ pytest -v testing/test_git.py::test_git_getdate_signed_commit
=========================================================================================== test session starts ============================================================================================
platform linux -- Python 3.8.10, pytest-7.0.0, pluggy-1.0.0 -- /home/ubuntu/setuptools_scm/venv/bin/python3
cachedir: .pytest_cache
setuptools version 60.8.1 from '/home/ubuntu/setuptools_scm/venv/lib/python3.8/site-packages/setuptools/__init__.py'
setuptools_scm version 6.4.3.dev10+gcc99e20 from '/home/ubuntu/setuptools_scm/src/setuptools_scm/__init__.py'
rootdir: /home/ubuntu/setuptools_scm, configfile: tox.ini
collected 1 item                                                                                                                                                                                           

testing/test_git.py::test_git_getdate_signed_commit FAILED                                                                                                                                           [100%]

================================================================================================= FAILURES =================================================================================================
______________________________________________________________________________________ test_git_getdate_signed_commit ______________________________________________________________________________________

signed_commit_wd = <WD /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/wd>

    @pytest.mark.issue("https://github.com/pypa/setuptools_scm/issues/548")
    def test_git_getdate_signed_commit(signed_commit_wd):
        today = date.today()
        signed_commit_wd.commit_testfile(signed=True)
        git_wd = git.GitWorkdir(os.fspath(signed_commit_wd.cwd))
>       assert git_wd.get_head_date() == today

testing/test_git.py:436: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/setuptools_scm/git.py:81: in get_head_date
    return datetime.strptime(date_part, r"%Y-%m-%d").date()
/usr/lib/python3.8/_strptime.py:568: in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

data_string = 'gpg: Signature made Mon Feb  7 14:41:13 2022 U', format = '%Y-%m-%d'

    def _strptime(data_string, format="%a %b %d %H:%M:%S %Y"):
        """Return a 2-tuple consisting of a time struct and an int containing
        the number of microseconds based on the input string and the
        format string."""
    
        for index, arg in enumerate([data_string, format]):
            if not isinstance(arg, str):
                msg = "strptime() argument {} must be str, not {}"
                raise TypeError(msg.format(index, type(arg)))
    
        global _TimeRE_cache, _regex_cache
        with _cache_lock:
            locale_time = _TimeRE_cache.locale_time
            if (_getlang() != locale_time.lang or
                time.tzname != locale_time.tzname or
                time.daylight != locale_time.daylight):
                _TimeRE_cache = TimeRE()
                _regex_cache.clear()
                locale_time = _TimeRE_cache.locale_time
            if len(_regex_cache) > _CACHE_MAX_SIZE:
                _regex_cache.clear()
            format_regex = _regex_cache.get(format)
            if not format_regex:
                try:
                    format_regex = _TimeRE_cache.compile(format)
                # KeyError raised when a bad format is found; can be specified as
                # \\, in which case it was a stray % but with a space after it
                except KeyError as err:
                    bad_directive = err.args[0]
                    if bad_directive == "\\":
                        bad_directive = "%"
                    del err
                    raise ValueError("'%s' is a bad directive in format '%s'" %
                                        (bad_directive, format)) from None
                # IndexError only occurs when the format string is "%"
                except IndexError:
                    raise ValueError("stray %% in format '%s'" % format) from None
                _regex_cache[format] = format_regex
        found = format_regex.match(data_string)
        if not found:
>           raise ValueError("time data %r does not match format %r" %
                             (data_string, format))
E           ValueError: time data 'gpg: Signature made Mon Feb  7 14:41:13 2022 U' does not match format '%Y-%m-%d'

/usr/lib/python3.8/_strptime.py:349: ValueError
------------------------------------------------------------------------------------------ Captured stderr setup -------------------------------------------------------------------------------------------
cmd 'git init'
 in /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/wd
out b'Initialized empty Git repository in /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/wd/.git/\n'
cmd 'git config user.email test@example.com'
 in /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/wd
cmd 'git config user.name "a test"'
 in /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/wd
cmd 'gpg --batch --generate-key /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/gpg_batch_params'
 in /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/wd
err b"gpg: keybox '/tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/pubring.kbx' created\ngpg: /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/trustdb.gpg: trustdb created\ngpg: key B5A97A7C6E04EEA4 marked as ultimately trusted\ngpg: directory '/tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/openpgp-revocs.d' created\ngpg: revocation certificate stored as '/tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/openpgp-revocs.d/803A90EF1466EC24ACC11420B5A97A7C6E04EEA4.rev'\n"
cmd 'git config log.showSignature true'
 in /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/wd
------------------------------------------------------------------------------------------- Captured stderr call -------------------------------------------------------------------------------------------
cmd 'git add .'
 in /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/wd
cmd 'git commit -S -m test-number-0'
 in /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/wd
out b'[master (root-commit) 24a0d67] test-number-0\n 1 file changed, 1 insertion(+)\n create mode 100644 test.txt\n'
cmd 'git log -n 1 HEAD --format=%cI'
 in /tmp/pytest-of-ubuntu/pytest-19/test_git_getdate_signed_commit0/wd
out b'gpg: Signature made Mon Feb  7 14:41:13 2022 UTC\ngpg:                using RSA key 803A90EF1466EC24ACC11420B5A97A7C6E04EEA4\ngpg:                issuer "test@example.com"\ngpg: checking the trustdb\ngpg: marginals needed: 3  completes needed: 1  trust model: pgp\ngpg: depth: 0  valid:   1  signed:   0  trust: 0-, 0q, 0n, 0m, 0f, 1u\ngpg: Good signature from "a test <test@example.com>" [ultimate]\n2022-02-07T14:41:13+00:00\n'
========================================================================================= short test summary info ==========================================================================================
FAILED testing/test_git.py::test_git_getdate_signed_commit - ValueError: time data 'gpg: Signature made Mon Feb  7 14:41:13 2022 U' does not match format '%Y-%m-%d'
============================================================================================ 1 failed in 1.38s =============================================================================================
```
</details>

<details><summary>Test skipping (due to missing <code>gpg</code>)</summary>

```
(venv) ubuntu@ephemeral:~/setuptools_scm$ pytest -v testing/test_git.py::test_git_getdate_signed_commit
=========================================================================================== test session starts ============================================================================================
platform linux -- Python 3.8.10, pytest-7.0.0, pluggy-1.0.0 -- /home/ubuntu/setuptools_scm/venv/bin/python3
cachedir: .pytest_cache
setuptools version 60.8.1 from '/home/ubuntu/setuptools_scm/venv/lib/python3.8/site-packages/setuptools/__init__.py'
setuptools_scm version 6.4.3.dev10+gcc99e20 from '/home/ubuntu/setuptools_scm/src/setuptools_scm/__init__.py'
rootdir: /home/ubuntu/setuptools_scm, configfile: tox.ini
collected 1 item                                                                                                                                                                                           

testing/test_git.py::test_git_getdate_signed_commit SKIPPED (gpg executable not found)                                                                                                               [100%]

============================================================================================ 1 skipped in 0.22s ============================================================================================
```
</details>

<details><summary>Test passing (due to #681)</summary>

```
(venv) ubuntu@ephemeral:~/setuptools_scm$ pytest -v testing/test_git.py::test_git_getdate_signed_commit
=========================================================================================== test session starts ============================================================================================
platform linux -- Python 3.8.10, pytest-7.0.0, pluggy-1.0.0 -- /home/ubuntu/setuptools_scm/venv/bin/python3
cachedir: .pytest_cache
setuptools version 60.8.1 from '/home/ubuntu/setuptools_scm/venv/lib/python3.8/site-packages/setuptools/__init__.py'
setuptools_scm version 6.4.3.dev10+gcc99e20 from '/home/ubuntu/setuptools_scm/src/setuptools_scm/__init__.py'
rootdir: /home/ubuntu/setuptools_scm, configfile: tox.ini
collected 1 item                                                                                                                                                                                           

testing/test_git.py::test_git_getdate_signed_commit PASSED                                                                                                                                           [100%]

============================================================================================ 1 passed in 1.54s =============================================================================================
```
</details>

puxlit committed:
Support signed commit commands in tests

puxlit committed:
Add test case for #548

puxlit committed:
Make implicit `"help"` program argument overridable in `has_command`

puxlit commented:
Hmm, okay it looks like the Windows builds are failing for the wrong reason (`gpg: can't connect to the agent: Invalid value passed to IPC`). I'll see if I can figure out what's going on.

puxlit commented:
Alright, I think I see the problem.

On Windows, we happen to be using the GnuPG that's bundled with Git for Windows.

```
DEBUG:    3+  >>>> Get-Command gpg


CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Application     gpg.exe                                            0.0.0.0    C:\Program Files\Git\usr\bin\gpg.exe
DEBUG:    4+  >>>> gpgconf --list-dirs

sysconfdir:/etc/gnupg
bindir:/usr/bin
libexecdir:/usr/lib/gnupg
libdir:/usr/lib/gnupg
datadir:/usr/share/gnupg
localedir:/usr/share/locale
socketdir:/c/Users/runneradmin/.gnupg
dirmngr-socket:/c/Users/runneradmin/.gnupg/S.dirmngr
agent-ssh-socket:/c/Users/runneradmin/.gnupg/S.gpg-agent.ssh
agent-extra-socket:/c/Users/runneradmin/.gnupg/S.gpg-agent.extra
agent-browser-socket:/c/Users/runneradmin/.gnupg/S.gpg-agent.browser
agent-socket:/c/Users/runneradmin/.gnupg/S.gpg-agent
homedir:/c/Users/runneradmin/.gnupg
DEBUG:    5+  >>>> gpg-connect-agent /bye

gpg-connect-agent: no running gpg-agent - starting '/usr/bin/gpg-agent'
gpg-connect-agent: failed to create temporary file '/c/Users/runneradmin/.gnupg/.#lk0x0000000800057d90.fv-az211-764.1821': No such file or directory
gpg-connect-agent: can't connect to the agent: No such file or directory
gpg-connect-agent: error sending standard options: No agent running
```

Thus, setting `GNUPGHOME` to a Windows-style path is no good.

```
DEBUG:    6+  >>>> $env:GNUPGHOME = Get-Location

DEBUG:    7+  >>>> gpgconf --list-dirs

sysconfdir:/etc/gnupg
bindir:/usr/bin
libexecdir:/usr/lib/gnupg
libdir:/usr/lib/gnupg
datadir:/usr/share/gnupg
localedir:/usr/share/locale
socketdir:D%3a\a\ephemeral-sandbox\ephemeral-sandbox
dirmngr-socket:D%3a\a\ephemeral-sandbox\ephemeral-sandbox/S.dirmngr
agent-ssh-socket:D%3a\a\ephemeral-sandbox\ephemeral-sandbox/S.gpg-agent.ssh
agent-extra-socket:D%3a\a\ephemeral-sandbox\ephemeral-sandbox/S.gpg-agent.extra
agent-browser-socket:D%3a\a\ephemeral-sandbox\ephemeral-sandbox/S.gpg-agent.browser
agent-socket:D%3a\a\ephemeral-sandbox\ephemeral-sandbox/S.gpg-agent
homedir:D%3a\a\ephemeral-sandbox\ephemeral-sandbox
DEBUG:    8+  >>>> gpg-connect-agent /bye

gpg-connect-agent: no running gpg-agent - starting '/usr/bin/gpg-agent'
gpg-connect-agent: waiting for the agent to come up ... (5s)
gpg-connect-agent: waiting for the agent to come up ... (4s)
gpg-connect-agent: waiting for the agent to come up ... (3s)
gpg-connect-agent: waiting for the agent to come up ... (2s)
gpg-connect-agent: waiting for the agent to come up ... (1s)
gpg-connect-agent: can't connect to the agent: Invalid value passed to IPC
gpg-connect-agent: error sending standard options: No agent running
```

Instead, setting it to a Unix-style path (via `cygpath`) appears to do the trick.

```
DEBUG:    9+  >>>> $env:GNUPGHOME = cygpath $(Get-Location)

DEBUG:    9+ $env:GNUPGHOME = cygpath $( >>>> Get-Location)

DEBUG:   10+  >>>> gpgconf --list-dirs

sysconfdir:/etc/gnupg
bindir:/usr/bin
libexecdir:/usr/lib/gnupg
libdir:/usr/lib/gnupg
datadir:/usr/share/gnupg
localedir:/usr/share/locale
socketdir:/d/a/ephemeral-sandbox/ephemeral-sandbox
dirmngr-socket:/d/a/ephemeral-sandbox/ephemeral-sandbox/S.dirmngr
agent-ssh-socket:/d/a/ephemeral-sandbox/ephemeral-sandbox/S.gpg-agent.ssh
agent-extra-socket:/d/a/ephemeral-sandbox/ephemeral-sandbox/S.gpg-agent.extra
agent-browser-socket:/d/a/ephemeral-sandbox/ephemeral-sandbox/S.gpg-agent.browser
agent-socket:/d/a/ephemeral-sandbox/ephemeral-sandbox/S.gpg-agent
homedir:/d/a/ephemeral-sandbox/ephemeral-sandbox
DEBUG:   11+  >>>> gpg-connect-agent /bye

gpg-connect-agent: no running gpg-agent - starting '/usr/bin/gpg-agent'
gpg-connect-agent: waiting for the agent to come up ... (5s)
gpg-connect-agent: connection to agent established
```

Alternatively, the [latest version](https://community.chocolatey.org/packages/gnupg/2.3.4) of GnuPG works with Windows-style paths. However, I can't seem to get `choco install gnupg -y` to run as a GitHub Actions workflow step without hanging. 😔

Still, I imagine installing a "Windows-compatible" version of GnuPG would be preferable, as that means the test wouldn't have to futz around with path style conversions?

RonnyPfannschmidt commented:
It's acceptable to leave the details for later as long as we are reasonably sure it works for users

puxlit committed:
Install GnuPG where necessary for test CI

puxlit commented:
I think I've figured it out. It's probably actions/virtual-environments#2876, where the installer fails with [`PATH env variable too big`](https://dev.gnupg.org/T5593). ([As of v2.3.4](https://dev.gnupg.org/source/gnupg/browse/master/build-aux/speedo/w32/g4wihelp.c;f74c65fd9bbb97a24b1a3bfe73706c6fcf95ad2a$973), it appears the limit is 2,047 bytes, whereas the current Windows runner's `PATH` length comes to 2,728 bytes.) I've updated the workflow with the workaround suggested in https://github.com/actions/virtual-environments/issues/2876#issuecomment-793653866. 🤞

RonnyPfannschmidt reviewed (approved):
Great work, will merge once it's green

puxlit committed:
Pick and configure an explicit `Key-Type`

If the default `Key-Type` happens to be an ECC, then `Key-Curve` would be a required parameter, so let's just pick RSA2048 and be done with it.

puxlit commented:
Oh dear, more unexpected Windows failures… 😂

Fixed `gpg: key generation failed: Unknown elliptic curve` in 96a22aa2ccfe50beeba713e8bec5f6694473134a. Fixed `gpg: invalid size of lockfile` in b4d7c786125ecb2f73fb43b247707db5dd00f50a. Tested this works in a Windows VM.

<details><summary>Test failing the <em>expected</em> way on Windows</summary>

```
(venv) PS C:\Users\Administrator\setuptools_scm> pytest -v testing/test_git.py::test_git_getdate_signed_commit
========================================================================================== test session starts ===========================================================================================
platform win32 -- Python 3.10.2, pytest-7.0.0, pluggy-1.0.0 -- C:\Users\Administrator\setuptools_scm\venv\Scripts\python.exe
cachedir: .pytest_cache
setuptools version 60.8.1 from 'C:\\Users\\Administrator\\setuptools_scm\\venv\\lib\\site-packages\\setuptools\\__init__.py'
setuptools_scm version 6.4.3.dev10+gced4242 from 'c:\\users\\administrator\\setuptools_scm\\src\\setuptools_scm\\__init__.py'
rootdir: C:\Users\Administrator\setuptools_scm, configfile: tox.ini
collected 1 item

testing/test_git.py::test_git_getdate_signed_commit FAILED                                                                                                                                          [100%]

================================================================================================ FAILURES ================================================================================================
_____________________________________________________________________________________ test_git_getdate_signed_commit _____________________________________________________________________________________

signed_commit_wd = <WD C:\Users\Administrator\AppData\Local\Temp\2\pytest-of-Administrator\pytest-14\test_git_getdate_signed_commit0\wd>

    @pytest.mark.issue("https://github.com/pypa/setuptools_scm/issues/548")
    def test_git_getdate_signed_commit(signed_commit_wd):
        today = date.today()
        signed_commit_wd.commit_testfile(signed=True)
        git_wd = git.GitWorkdir(os.fspath(signed_commit_wd.cwd))
>       assert git_wd.get_head_date() == today

testing\test_git.py:437:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src\setuptools_scm\git.py:81: in get_head_date
    return datetime.strptime(date_part, r"%Y-%m-%d").date()
C:\Python310\lib\_strptime.py:568: in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

data_string = 'gpg: Signature made 2/9/2022 11:10:16 AM Coordinated Universal ', format = '%Y-%m-%d'

    def _strptime(data_string, format="%a %b %d %H:%M:%S %Y"):
        """Return a 2-tuple consisting of a time struct and an int containing
        the number of microseconds based on the input string and the
        format string."""

        for index, arg in enumerate([data_string, format]):
            if not isinstance(arg, str):
                msg = "strptime() argument {} must be str, not {}"
                raise TypeError(msg.format(index, type(arg)))

        global _TimeRE_cache, _regex_cache
        with _cache_lock:
            locale_time = _TimeRE_cache.locale_time
            if (_getlang() != locale_time.lang or
                time.tzname != locale_time.tzname or
                time.daylight != locale_time.daylight):
                _TimeRE_cache = TimeRE()
                _regex_cache.clear()
                locale_time = _TimeRE_cache.locale_time
            if len(_regex_cache) > _CACHE_MAX_SIZE:
                _regex_cache.clear()
            format_regex = _regex_cache.get(format)
            if not format_regex:
                try:
                    format_regex = _TimeRE_cache.compile(format)
                # KeyError raised when a bad format is found; can be specified as
                # \\, in which case it was a stray % but with a space after it
                except KeyError as err:
                    bad_directive = err.args[0]
                    if bad_directive == "\\":
                        bad_directive = "%"
                    del err
                    raise ValueError("'%s' is a bad directive in format '%s'" %
                                        (bad_directive, format)) from None
                # IndexError only occurs when the format string is "%"
                except IndexError:
                    raise ValueError("stray %% in format '%s'" % format) from None
                _regex_cache[format] = format_regex
        found = format_regex.match(data_string)
        if not found:
>           raise ValueError("time data %r does not match format %r" %
                             (data_string, format))
E           ValueError: time data 'gpg: Signature made 2/9/2022 11:10:16 AM Coordinated Universal ' does not match format '%Y-%m-%d'

C:\Python310\lib\_strptime.py:349: ValueError
----------------------------------------------------------------------------------------- Captured stderr setup ------------------------------------------------------------------------------------------
cmd 'git init'
 in C:\Users\Administrator\AppData\Local\Temp\2\pytest-of-Administrator\pytest-14\test_git_getdate_signed_commit0\wd
out b'Initialized empty Git repository in C:/Users/Administrator/AppData/Local/Temp/2/pytest-of-Administrator/pytest-14/test_git_getdate_signed_commit0/wd/.git/\n'
cmd 'git config user.email test@example.com'
 in C:\Users\Administrator\AppData\Local\Temp\2\pytest-of-Administrator\pytest-14\test_git_getdate_signed_commit0\wd
cmd 'git config user.name "a test"'
 in C:\Users\Administrator\AppData\Local\Temp\2\pytest-of-Administrator\pytest-14\test_git_getdate_signed_commit0\wd
cmd 'gpg --batch --generate-key C:\\Users\\Administrator\\AppData\\Local\\Temp\\2\\pytest-of-Administrator\\pytest-14\\test_git_getdate_signed_commit0\\gpg_batch_params'
 in C:\Users\Administrator\AppData\Local\Temp\2\pytest-of-Administrator\pytest-14\test_git_getdate_signed_commit0\wd
err b"gpg: keybox 'C:\\\\Users\\\\Administrator\\\\AppData\\\\Local\\\\Temp\\\\2\\\\pytest-of-Administrator\\\\pytest-14\\\\test_git_getdate_signed_commit0\\\\pubring.kbx' created\r\ngpg: C:\\\\Users\\\\Administrator\\\\AppData\\\\Local\\\\Temp\\\\2\\\\pytest-of-Administrator\\\\pytest-14\\\\test_git_getdate_signed_commit0\\\\trustdb.gpg: trustdb created\r\ngpg: directory 'C:\\\\Users\\\\Administrator\\\\AppData\\\\Local\\\\Temp\\\\2\\\\pytest-of-Administrator\\\\pytest-14\\\\test_git_getdate_signed_commit0\\\\openpgp-revocs.d' created\r\ngpg: revocation certificate stored as 'C:\\\\Users\\\\Administrator\\\\AppData\\\\Local\\\\Temp\\\\2\\\\pytest-of-Administrator\\\\pytest-14\\\\test_git_getdate_signed_commit0\\\\openpgp-revocs.d\\\\4C8970E10C8D4525CA16E4BAA7DD296C10BB5603.rev'\r\n"
cmd 'git config log.showSignature true'
 in C:\Users\Administrator\AppData\Local\Temp\2\pytest-of-Administrator\pytest-14\test_git_getdate_signed_commit0\wd
------------------------------------------------------------------------------------------ Captured stderr call ------------------------------------------------------------------------------------------
cmd 'git add .'
 in C:\Users\Administrator\AppData\Local\Temp\2\pytest-of-Administrator\pytest-14\test_git_getdate_signed_commit0\wd
cmd 'git commit -S -m test-number-0'
 in C:\Users\Administrator\AppData\Local\Temp\2\pytest-of-Administrator\pytest-14\test_git_getdate_signed_commit0\wd
out b'[master (root-commit) 11d4326] test-number-0\n 1 file changed, 1 insertion(+)\n create mode 100644 test.txt\n'
cmd 'git log -n 1 HEAD --format=%cI'
 in C:\Users\Administrator\AppData\Local\Temp\2\pytest-of-Administrator\pytest-14\test_git_getdate_signed_commit0\wd
out b'gpg: Signature made 2/9/2022 11:10:16 AM Coordinated Universal Time\r\ngpg:                using RSA key 4C8970E10C8D4525CA16E4BAA7DD296C10BB5603\r\ngpg:                issuer "test@example.com"\r\ngpg: checking the trustdb\r\ngpg: marginals needed: 3  completes needed: 1  trust model: pgp\r\ngpg: depth: 0  valid:   1  signed:   0  trust: 0-, 0q, 0n, 0m, 0f, 1u\r\ngpg: Good signature from "a test <test@example.com>" [ultimate]\r\n2022-02-09T11:10:16+00:00\n'
======================================================================================== short test summary info =========================================================================================
FAILED testing/test_git.py::test_git_getdate_signed_commit - ValueError: time data 'gpg: Signature made 2/9/2022 11:10:16 AM Coordinated Universal ' does not match format '%Y-%m-%d'
=========================================================================================== 1 failed in 3.97s ============================================================================================
```
</details>

So… hopefully third time's the charm. If the Windows builds fail the right way, I'll merge this branch into #681, to confirm that those changes fix these broken tests.

puxlit committed:
Explicitly set `gpg.program` to avoid <https://dev.gnupg.org/T5504>

puxlit commented:
```
git: D:\a\_temp\239dd12a-4daa-4d71-9a0f-debce348709a.ps1:6
Line |
   6 |  git config --system gpg.program "C:\Program Files (x86)\gnupg\bin\gpg …
     |  ~~~
     | The term 'git' is not recognized as a name of a cmdlet, function, script file, or executable program.
     | Check the spelling of the name, or if a path was included, verify that the path is correct and try
     | again.
```

… okay, _fourth_ time's the charm. (I forgot to extend the minimal `PATH` environment variable to include the path to `git`. 🤦)

Thanks for your patience!

RonnyPfannschmidt commented:
Thanks for the grit on this details plain.😁

puxlit commented:
Okay, at last! The new test is consistently failing the expected way! Merging into #681…

<dl>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125669511?check_suite_focus=true#step:9:91">windows-latest - Python 3.6</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made 2/9/2022 1:27:56 PM Coordinated Universal ' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125669578?check_suite_focus=true#step:9:92">ubuntu-latest - Python 3.6</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made Wed Feb  9 13:24:39 2022 U' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125669642?check_suite_focus=true#step:9:91">windows-latest - Python 3.7</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made 2/9/2022 1:29:45 PM Coordinated Universal ' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125669709?check_suite_focus=true#step:9:92">ubuntu-latest - Python 3.7</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made Wed Feb  9 13:28:03 2022 U' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125669768?check_suite_focus=true#step:9:91">windows-latest - Python 3.8</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made 2/9/2022 1:32:34 PM Coordinated Universal ' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125669842?check_suite_focus=true#step:9:92">ubuntu-latest - Python 3.8</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made Wed Feb  9 13:30:06 2022 U' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125669910?check_suite_focus=true#step:9:91">windows-latest - Python 3.9</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made 2/9/2022 1:33:37 PM Coordinated Universal ' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125669971?check_suite_focus=true#step:9:92">ubuntu-latest - Python 3.9</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made Wed Feb  9 13:29:39 2022 U' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125670058?check_suite_focus=true#step:9:91">windows-latest - Python 3.10</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made 2/9/2022 1:27:30 PM Coordinated Universal ' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125670170?check_suite_focus=true#step:9:92">ubuntu-latest - Python 3.10</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made Wed Feb  9 13:25:19 2022 U' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125670255?check_suite_focus=true#step:9:93">windows-latest - Python pypy3</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made 2/9/2022 1:33:28 PM Coordinated Universal ' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125670320?check_suite_focus=true#step:9:93">ubuntu-latest - Python pypy3</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made Wed Feb  9 13:36:18 2022 U' does not match format '%Y-%m-%d'</code></dd>
<dt><a href="https://github.com/pypa/setuptools_scm/runs/5125670393?check_suite_focus=true#step:9:91">windows-latest - Python msys2</a></dt>
<dd><code>E           ValueError: time data 'gpg: Signature made 2/9/2022 1:37:50 PM Coordinated Universal ' does not match format '%Y-%m-%d'</code></dd>
</dl>
